web3/_utils/async_tools.py

```diff
@@ -0,0 +1,84 @@
+import asyncio
+from concurrent.futures import (
+    ThreadPoolExecutor,
+)
+from functools import (
+    singledispatch,
+    wraps,
+)
+import inspect
+import types
+from typing import (
+    Any,
+    Callable,
+    Generator,
+)
+
+_DEFAULT_POOL = ThreadPoolExecutor()
+
+
+def threadpool(f, executor=None):
+    @wraps(f)
+    def wrap(*args, **kwargs):
+        return (executor or _DEFAULT_POOL).submit(f, *args, **kwargs)
+    return wrap
+
+
+@singledispatch
+@threadpool
+def run_in_new_loop_in_thread(co: Any, args, kwargs):
+    raise TypeError('Called with unsupported argument: {}'.format(co))
+
+
+@run_in_new_loop_in_thread.register(asyncio.Future)
+@run_in_new_loop_in_thread.register(types.GeneratorType)
+@run_in_new_loop_in_thread.register(types.CoroutineType)
+@threadpool
+def run_co_in_new_loop_in_thread(co: Generator[Any, None, Any]) -> Any:
+    loop = asyncio.new_event_loop()
+    return loop.run_until_complete(co)
+
+
+@run_in_new_loop_in_thread.register(types.FunctionType)
+@run_in_new_loop_in_thread.register(types.MethodType)
+@threadpool
+def run_f_in_new_loop_in_thread(f: Callable[..., Any]) -> Callable[..., Any]:
+    @wraps(f)
+    def run(*args, **kwargs):
+        loop = asyncio.new_event_loop()
+        return loop.run_until_complete(f(*args, **kwargs))
+    return run
+
+
+@singledispatch
+def sync(co: Any):
+    raise TypeError('Called with unsupported argument: {}'.format(co))
+
+
+@sync.register(asyncio.Future)
+@sync.register(types.GeneratorType)
+@sync.register(types.CoroutineType)
+def sync_co(co: Generator[Any, None, Any]) -> Any:
+    if not inspect.isawaitable(co):
```

Comment:
It'd be good to have some docstrings in this module explaining the "what" and "why" for each of these different pieces of functionality because it isn't intuitive why we need these different pieces of logic.